File: src/main/java/com/lessmarkup/engine/filesystem/ResourceCacheImpl.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.filesystem;

import com.google.inject.Inject;
import com.lessmarkup.dataobjects.SiteCustomization;
import com.lessmarkup.framework.helpers.LoggingHelper;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.cache.AbstractCacheHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.module.Implements;
import com.lessmarkup.interfaces.module.ModuleConfiguration;
import com.lessmarkup.interfaces.module.ModuleProvider;
import com.lessmarkup.interfaces.system.ResourceCache;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.OptionalLong;
import java.util.logging.Level;

import com.samskivert.mustache.Mustache;
import com.samskivert.mustache.Template;

@Implements(ResourceCache.class)
class ResourceCacheImpl extends AbstractCacheHandler implements ResourceCache {

    private final ModuleProvider moduleProvider;
    private final DomainModelProvider domainModelProvider;
    private final DataCache dataCache;
    private final Map<String, ResourceReference> resources = new HashMap<>();
    private Mustache.Compiler compiler;

    @Inject
    public ResourceCacheImpl(ModuleProvider moduleProvider, DomainModelProvider domainModelProvider, DataCache dataCache) {
        super(null);
        this.moduleProvider = moduleProvider;
        this.domainModelProvider = domainModelProvider;
        this.dataCache = dataCache;
    }

    private static boolean isExtensionSupported(String extension) {
        switch (extension) {
            case "jpg":
            case "gif":
            case "png":
            case "html":
            case "xml":
            case "js":
            case "css":
            case "eot":
            case "svg":
            case "ttf":
            case "woff":
            case "woff2":
            case "map":
            case "ts":
                return true;
        }

        return false;
    }

    private static String extractExtension(String path) {
        int pos = path.lastIndexOf('.');
        if (pos <= 0) {
            return null;
        }

        return path.substring(pos+1).toLowerCase();
    }

    @Override
    public void initialize(OptionalLong objectId) {

        for (ModuleConfiguration module : moduleProvider.getModules()){
            for (String element : module.getElements()) {

                String extension = extractExtension(element);

                if (extension != null && isExtensionSupported(extension)) {
                    addResource(module, element, extension);
                }
            }
        }
        
        loadDatabaseResources();
        
        this.compiler = TemplateContext.createCompiler(this);
    }
    
    private void loadDatabaseResources() {
        if (RequestContextHolder.getContext().getEngineConfiguration().isCustomizationsDisabled()) {
            return;
        }
        
        try (DomainModel domainModel = domainModelProvider.create()) {
            for (SiteCustomization record : domainModel.query().from(SiteCustomization.class).toList(SiteCustomization.class)) {
                String recordPath = record.getPath().toLowerCase();
                
                ResourceReference reference;
                
                if (record.isAppend()) {
                    reference = resources.get(recordPath);
                    if (reference != null) {
                        byte[] binary = new byte[reference.getBinary().length + record.getBody().length];
                        System.arraycopy(reference.getBinary(), 0, binary, 0, reference.getBinary().length);
                        System.arraycopy(record.getBody(), 0, binary, reference.getBinary().length, record.getBody().length);
                        reference.setBinary(binary);
                        continue;
                    }
                }
                
                reference = new ResourceReference();
                reference.setRecordId(record.getId());
                reference.setBinary(record.getBody());
                
                resources.put(recordPath, reference);
            }
        }
    }
    
    private void addResource(ModuleConfiguration module, String path, String extension) {
        ResourceReference reference = new ResourceReference();
        reference.setModule(module);
        reference.setPath(path);
        reference.setExtension(extension);
        resources.put(path, reference);
    }
    
    @Override
    public boolean resourceExists(String path) {
        return resources.containsKey(path);
    }
    
    ResourceReference loadResource(String path) {
        ResourceReference resource = resources.get(path);
        if (resource == null) {
            return null;
        }
        
        if (resource.getBinary() != null) { 
            return resource;
        }
        
        //noinspection SynchronizationOnLocalVariableOrMethodParameter
        synchronized(resource) {
            try {
                resource.setBinary(resource.getModule().getResourceAsBytes(path));
            } catch (IOException e) {
                LoggingHelper.getLogger(getClass()).log(Level.SEVERE, null, e);
            }

            if (resource.getBinary() != null) {
                if ("html".equals(resource.getExtension())) {
                    String body = StringHelper.binaryToString(resource.getBinary());
                    if (body.contains("[[")) {
                        Template template = compiler.compile(body);
                        resource.setTemplate(template);
                    }
                }
            }
        }
        return resource;
    }
    
    @Override
    public byte[] readBytes(String path) {
        ResourceReference resource = loadResource(path);
        if (resource == null || resource.getBinary() == null) {
            return null;
        }
        return resource.getBinary();
    }

    @Override
    public String readText(String path) {
        ResourceReference resource = loadResource(path);
        if (resource == null || resource.getBinary() == null) {
            return null;
        }
        return StringHelper.binaryToString(resource.getBinary());
    }
    
    String parseText(ResourceReference reference) {
        if (reference.getTemplate() == null) {
            return StringHelper.binaryToString(reference.getBinary());
        }
        TemplateContext context = new TemplateContext(dataCache);
        return reference.getTemplate().execute(context);
    }

    @Override
    public String parseText(String path) {
        ResourceReference reference = loadResource(path);
        if (reference == null || reference.getBinary() == null) {
            return null;
        }
        return parseText(reference);
    }
}

File: src/main/scala/com/lessmarkup/engine/filesystem/ResourceCacheImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.filesystem

import java.io.IOException
import java.util.logging.Level

import com.google.inject.Inject
import com.lessmarkup.dataobjects.SiteCustomization
import com.lessmarkup.framework.helpers.{LoggingHelper, StringHelper}
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.Implements
import com.lessmarkup.interfaces.cache.{AbstractCacheHandler, DataCache}
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider}
import com.lessmarkup.interfaces.module.{ModuleConfiguration, ModuleProvider}
import com.lessmarkup.interfaces.system.ResourceCache
import com.samskivert.mustache.{Mustache, Template}

import scala.collection.JavaConversions._

object ResourceCacheImpl {
  private def isExtensionSupported(extension: String): Boolean = {
    extension match {
      case "jpg" => true
      case "gif" => true
      case "png" => true
      case "html" => true
      case "xml" => true
      case "js" => true
      case "css" => true
      case "eot" => true
      case "svg" => true
      case "ttf" => true
      case "woff" => true
      case "woff2" => true
      case "map" => true
      case "ts" => true
      case _ => false
    }
  }

  private def extractExtension(path: String): Option[String] = {
    val pos: Int = path.lastIndexOf('.')
    if (pos <= 0) {
      None
    } else {
      Option(path.substring(pos + 1).toLowerCase)
    }
  }
}

@Implements(classOf[ResourceCache])
class ResourceCacheImpl @Inject() (moduleProvider: ModuleProvider, domainModelProvider: DomainModelProvider, dataCache: DataCache) extends AbstractCacheHandler(null) with ResourceCache {
  private val resources = loadDatabaseResources

  private val compiler: Mustache.Compiler = TemplateContext.createCompiler(this)

  private def loadModulesResources: Map[String, ResourceReference] = {

    moduleProvider.getModules
      .flatten(m => m.getElements.map(e => (m, e, ResourceCacheImpl.extractExtension(e))))
      .filter(e => e._3.isDefined && ResourceCacheImpl.isExtensionSupported(e._3.get))
      .map(e => (e._2, createResource(e._1, e._2, e._3.get))).toMap
  }

  private def loadDatabaseResources: Map[String, ResourceReference] = {

    val modulesResources = loadModulesResources

    if (RequestContextHolder.getContext.getEngineConfiguration.isCustomizationsDisabled) {
      return modulesResources
    }

    val domainModel: DomainModel = domainModelProvider.create
    try {

      domainModel.query.from(classOf[SiteCustomization]).toList(classOf[SiteCustomization]).map(record => {
        val recordPath: String = record.path
        val reference: Option[ResourceReference] = if (record.append) modulesResources.get(recordPath) else None
        if (record.append && reference.isDefined) {
          val binaryLength = if (reference.get.binary.isDefined) reference.get.binary.get.length else 0
          val binary: Array[Byte] = new Array[Byte](binaryLength + record.body.length)
          if (reference.get.binary.isDefined) {
            System.arraycopy(reference.get.binary.get, 0, binary, 0, binaryLength)
          }
          System.arraycopy(record.body, 0, binary, binaryLength, record.body.length)
          reference.get.binary = Option(binary)
          None
        } else {
          Option(new ResourceReference(
            path = record.path,
            recordId = record.id,
            binary = Option(record.body)
          ))
        }
      }).filter(_.isDefined)
        .map(r => (r.get.path, r.get))
        .toMap ++ modulesResources

    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  private def createResource(module: ModuleConfiguration, path: String, extension: String): ResourceReference = {
    new ResourceReference(
      module = Option(module),
      path = path,
      extension = Option(extension))
  }

  def resourceExists(path: String): Boolean = {
    resources.containsKey(path)
  }

  private def loadResource(path: String): Option[ResourceReference] = {
    val resourceOption: Option[ResourceReference] = resources.get(path)
    if (resourceOption.isEmpty) {
      return resourceOption
    }

    val resource = resourceOption.get

    if (resource.binary.isDefined || resource.module.isEmpty) {
      return resourceOption
    }

    resource synchronized {
      try {
        resource.binary = Option(resource.module.get.getResourceAsBytes(path))
      }
      catch {
        case e: IOException =>
          LoggingHelper.getLogger(getClass).log(Level.SEVERE, null, e)
      }
      if (resource.binary.isDefined) {
        if (resource.extension.isDefined && "html" == resource.extension.get) {
          val body: String = StringHelper.binaryToString(resource.binary.get)
          if (body.contains("[[")) {
            val template: Template = compiler.compile(body)
            resource.template = Option(template)
          }
        }
      }
    }

    resourceOption
  }

  def readBytes(path: String): Option[Array[Byte]] = {
    val resource: Option[ResourceReference] = loadResource(path)
    if (resource.isEmpty || resource.get.binary.isEmpty) {
      None
    } else {
      resource.get.binary
    }
  }

  def readText(path: String): Option[String] = {
    val resource = loadResource(path)
    if (resource.isEmpty || resource.get.binary.isEmpty) {
      None
    } else {
      Option(StringHelper.binaryToString(resource.get.binary.get))
    }
  }

  private def parseText(reference: ResourceReference): String = {
    if (reference.template.isEmpty) {
      StringHelper.binaryToString(reference.binary.get)
    } else {
      val context: TemplateContext = new TemplateContext(dataCache)
      reference.template.get.execute(context)
    }
  }

  def parseText(path: String): Option[String] = {
    val reference = loadResource(path)
    if (reference.isEmpty || reference.get.binary.isEmpty) {
      None
    } else {
      Option(parseText(reference.get))
    }
  }
}
